File: src/main/java/com/example/desafio/api/services/BrandService.java
package com.example.desafio.api.services;

import com.example.desafio.api.exceptions.CarExceptions;
import com.example.desafio.api.models.dtos.BrandDto;
import com.example.desafio.api.models.dtos.BrandRequestDto;
import com.example.desafio.api.models.entitys.BrandEntity;
import com.example.desafio.api.models.repositorys.BrandRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class BrandService {

    private final BrandRepository brandRepository;

    @Autowired
    public BrandService(BrandRepository brandRepository) {
        this.brandRepository = brandRepository;
    }

    /**
     * Método responsável por ATUALIZAR uma Marca
     *
     * @param id
     * @param brandDto
     * @return ResponseEntity<ApiResponseService>
     */
    public ResponseEntity<ApiResponseService> updateBrand(Long id, BrandRequestDto brandRequestDto) {
        try {

            BrandEntity brandEntity = brandRepository.findById(id).orElseThrow(() -> new CarExceptions("Marca não encontrada"));

            brandEntity.setNameBrand(brandRequestDto.getNameBrand());

            brandRepository.save(brandEntity);

            return ApiResponseService.createSuccessResponse("Marca atualizada com sucesso", null);

        } catch (Exception e) {
            return ApiResponseService.createErrorResponse("Erro ao atualizar a marca: " + e.getMessage());
        }
    }

    /**
     * Método responsável para BUSCAR todas as marcas
     *
     * @return ResponseEntity<ApiResponseService>
     */
    public ResponseEntity<ApiResponseService> findAll() {
        try {

            List<BrandEntity> allBrands = brandRepository.findAll();

            // Crio uma lista de DTO que vou preencher com os valores do banco
            List<BrandDto> brandDtos = new ArrayList<>();

            for (BrandEntity brand : allBrands) {

                // Crio uma instancia de DTO seto os valores, logo depois adiciono eles na lista
                BrandDto brandDto = new BrandDto();
                brandDto.setId(brand.getId());
                brandDto.setDeleted(brand.getDeleted());
                brandDto.setNameBrand(brand.getNameBrand());
                brandDtos.add(brandDto);
            }

            return ApiResponseService.createSuccessResponse("Marcas encontradas com sucesso", brandDtos);
        } catch (Exception e) {
            return ApiResponseService.createErrorResponse("Erro ao buscar as marcas: " + e.getMessage());
        }
    }

    /**
     * Método responsável por DELETAR uma marca e seus modelos associados
     *
     * @param id
     * @return ResponseEntity<ApiResponseService>
     */
    public ResponseEntity<ApiResponseService> deleteBrand(Long id) {
        try {

            BrandEntity brandEntity = brandRepository.findById(id).orElseThrow(() -> new CarExceptions("Marca não encontrada"));

            brandEntity.setDeleted(true);

            // Deleta todos os modelos associados a marca
            brandEntity.getModelCars().forEach(modelCar -> modelCar.setDeleted(true));

            // Deleta todos os carros associados aos modelos deletados
            brandEntity.getModelCars().forEach(modelCar -> modelCar.getCars().forEach(car -> car.setDeleted(true)));

            brandRepository.save(brandEntity);

            return ApiResponseService.createSuccessResponse("Marca deletada com sucesso", null);

        } catch (Exception e) {
            return ApiResponseService.createErrorResponse("Erro ao deletar a marca: " + e.getMessage());
        }
    }
}

File: src/main/kotlin/com/example/desafio/api/services/BrandService.kt
package com.example.desafio.api.services

import com.example.desafio.api.exceptions.CarExceptions
import com.example.desafio.api.models.dtos.BrandDto
import com.example.desafio.api.models.dtos.BrandRequestDto
import com.example.desafio.api.models.repositorys.BrandRepository
import com.example.desafio.api.services.ApiResponseService.Companion.createErrorResponse
import com.example.desafio.api.services.ApiResponseService.Companion.createSuccessResponse
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service

@Service
class BrandService @Autowired constructor(private val brandRepository: BrandRepository) {

    /**
     * Método responsável por ATUALIZAR marca
     *
     * @param id
     * @param brandRequestDto
     * @return ResponseEntity<ApiResponseService>
    </ApiResponseService> */
    fun updateBrand(id: Long, brandRequestDto: BrandRequestDto): ResponseEntity<ApiResponseService> {

        return try {

            val brandEntity = brandRepository.findById(id).orElseThrow { CarExceptions("Marca não encontrada") }!!

            brandEntity.nameBrand = brandRequestDto.nameBrand

            brandRepository.save(brandEntity)

            createSuccessResponse("Marca atualizada com sucesso", null)
        } catch (e: Exception) {
            createErrorResponse("Erro ao atualizar a marca: " + e.message)
        }
    }

    /**
     * Método responsável por BUSCAR todas as marcas
     *
     * @return ResponseEntity<ApiResponseService>
    </ApiResponseService> */
    fun findAll(): ResponseEntity<ApiResponseService> {
        return try {
            val allBrands = brandRepository.findAll()

            // Crio uma lista de DTO que vou preencher com os valores do banco
            val brandDtos: MutableList<BrandDto> = ArrayList()
            for (brand in allBrands) {

                // Crio uma instância de DTO seto os valores, logo após, adiciono eles na lista
                val brandDto = BrandDto()

                brandDto.id = brand?.id
                brandDto.deleted = brand?.deleted
                brandDto.nameBrand = brand?.nameBrand

                brandDtos.add(brandDto)
            }

            createSuccessResponse("Marcas encontradas com sucesso", brandDtos)
        } catch (e: Exception) {
            createErrorResponse("Erro ao buscar as marcas: " + e.message)
        }
    }

    /**
     * Método responsável por DELETAR uma marca
     *
     * @param id
     * @return ResponseEntity<ApiResponseService>
    </ApiResponseService> */
    fun deleteBrand(id: Long): ResponseEntity<ApiResponseService> {
        return try {
            val brandEntity = brandRepository.findById(id).orElseThrow { CarExceptions("Marca não encontrada") }!!

            // Deleta a marca
            brandEntity.deleted = true

            // Deleta todos os modelos associados a marca
            brandEntity.modelCars?.forEach { modelCar ->
                modelCar.deleted = true
            }

            // Deleta todos os carros associados aos modelos deletados
            brandEntity.modelCars?.forEach { modelCarEntity ->
                modelCarEntity.cars?.forEach { car ->
                    car.deleted = true
                }
            }

            brandRepository.save(brandEntity)

            createSuccessResponse("Marca deletada com sucesso", null)
        } catch (e: Exception) {
            createErrorResponse("Erro ao deletar a marca: " + e.message)
        }
    }
}